### ISSUE TITLE
Distributing TFP

### ISSUE THREAD

**Issue by EliseJ** [tensorflow/probability#75]
Is it possible to implement data parallelism in an example such as 
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/bayesian_neural_network.py
?
Recently I simply edited this with Horovod and it runs although I am not convinced it is averaging gradients.
Does the max ELBO calculation need to be serial? 
Is there anything in the use of the distributions which prevents adata parallel run?

Thanks in advance for your help!

**Comment by dustinvtran** [comment]
I'm not familiar with Horovod. If you run the example with, e.g., multi-GPUs, it will implement data parallelism by default. The implementations should be agnostic. (We also have a TPU version of the same example but haven't shared it yet.)

Because TFP is just a chain of TensorFlow ops, are there any Horovod diagnostics to check if data parallelism is implemented correctly?

**Comment by EliseJ** [comment]
Thanks for the quick reply. Yes, I can profile the code and check the Timeline. Right now it runs but it is just super slow (much slower than serial).

**Comment by dustinvtran** [comment]
Excellent. Also to clarify, what does serial mean in this context?

**Comment by EliseJ** [comment]
Running the mini batches sequentially.

**Comment by AngelBerihuete** [comment]
Hi @EliseJ @dustinvtran 

I'm old user of Edward library, and recently I'm trying to switch to Edward2. I'm searching for TFP/Edward2 examples using distributed training as ```tf.distribute.MirroredStrategy```...but I can't find a simple and good example explaining how to do this.
Please, could you point me out to any simple example on how to use  tf.distribute.MirroredStrategy using TFP or Edward2?

A simple example would be
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

import tensorflow as tf
from tensorflow_probability import edward2 as ed


def build_toy_dataset(N, w, noise_std=0.25):
    D = len(w)
    x = np.random.randn(N, D)
    y = np.dot(x, w) + np.random.normal(0, noise_std, size=N)
    return x.astype(np.float32), y.astype(np.float32)


# ed.set_seed(42)
N = 4000  # number of data points
D = 10  # number of features
w_true = np.random.randn(D)

X_train, y_train = build_toy_dataset(N, w_true)
X_test, y_test = build_toy_dataset(N, w_true)


def regression_model(regressors):
    w = ed.Normal(
        loc=tf.zeros(regressors.shape[1]),
        scale=tf.ones(regressors.shape[1]),
        name="w")
    b = ed.Normal(loc=0., scale=1., name="b")
    y = ed.Normal(
        loc=tf.tensordot(regressors, w, [[1], [0]]) + b,
        scale=tf.ones(N), name="y")
    return y


log_joint = ed.make_log_joint_fn(regression_model)

y = regression_model(X_train)

# with tf.Session() as sess:  # or, e.g., tf.train.MonitoredSession()
#     y_train2 = sess.run(y)


tf.reset_default_graph()

w = tf.Variable(np.ones([X_train.shape[1]]), dtype=tf.float32)
b = tf.Variable(1., dtype=tf.float32)


def target(w, b):
    """Unnormalized target density as a function of the parameters."""
    return log_joint(regressors=X_train, w=w, b=b, y=y_train)


energy = -target(w, b)

optimizer = tf.train.AdamOptimizer(learning_rate=0.05)
train = optimizer.minimize(energy)

init = tf.global_variables_initializer()

t = []

num_epochs = 2000

with tf.Session() as sess:
    sess.run(init)

    for i in range(num_epochs):
        sess.run(train)
        if i % 5 == 0:
            cE, cw, cb = sess.run([energy, w, b])
            t.append(cE)

    w_inferred_map = sess.run(w)
    b_inferred_map = sess.run(b)


print("true w:")
print(w_true)

print("\nInferred w:")
print(w_inferred_map)
```
